### PR TITLE
fix immutability for ARC4 struct and address types

### PR DESCRIPTION
**Pull request by daniel-makerx**
BREAKING CHANGE: modifying an `algopy.arc4.Address` will now raise an error

### CODE REVIEW

**Review by achidlow — DISMISSED** [review]

**Review by achidlow — DISMISSED** [review]

File: src/puya/awst/validation/immutable.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from puya import log
+from puya.awst import nodes as awst_nodes
+from puya.awst.awst_traverser import AWSTTraverser
+
+logger = log.get_logger(__name__)
+
+
+class ImmutableValidator(AWSTTraverser):
```
Comment (tristanmenzel) [inline src/puya/awst/validation/immutable.py:8]: Should this include mutating nodes like `ArrayExtend` and `ArrayPop`?

Comment (daniel-makerx) [reply src/puya/awst/validation/immutable.py:8]: Good point, initially this issue was addressing arc4.Address, which those nodes don't operate on. But we should still check regardless.

**Comment by daniel-makerx** [comment]
Have found additional semantic compatibility issues around mutable arc4 types, which may require more breaking changes so converting to draft while addressing those

File: src/puyapy/awst_build/eb/arc4/_base.py [hunk 2]
```diff
@@ -138,9 +138,8 @@ class _ARC4ArrayExpressionBuilder(BytesBackedInstanceExpressionBuilder[pytypes.A
     @typing.override
     def iterate(self) -> Expression:
         if not self.pytype.items.wtype.immutable:
-            logger.error(
-                "cannot directly iterate an ARC4 array of mutable objects,"
-                " construct a for-loop over the indexes via urange(<array>.length) instead",
+            logger.info(
+                "use `algopy.urange(<array>.length)` to iterate by index",
```
Comment (achidlow) [inline src/puyapy/awst_build/eb/arc4/_base.py:142]: why remove the fact that it's due to mutable values from the error message here?

Comment (achidlow) [reply src/puyapy/awst_build/eb/arc4/_base.py:142]: oh I see - maybe add a code comment that this condition creates an error during validation in puya core

Comment (daniel-makerx) [reply src/puyapy/awst_build/eb/arc4/_base.py:145]: Added comment

**Review by achidlow — APPROVED** [review]

**Pull request merged**